aws: retrieve instance metadata from IMDS

snir911 opened the pull request:
if not explicitly set, retrieve subnet-id, region and key-name from
AWS Instance Metadata Service

Fixes: #315
Signed-off-by: Snir Sheriber <ssheribe@redhat.com>

(we may fetch also SGs later, this was in mind when created)

to test on aws: quay.io/snir/cloud-api-adaptor-aws:315

bpradipt commented:
@snir911 can you please split the entrypoint changes into a new PR

snir911 commented:
removed, will open the PR later

bpradipt commented on entrypoint.sh:
```diff
@@ -26,18 +26,16 @@ test_vars() {
+[[ "${PODVM_INSTANCE_TYPE}" ]] && optionals+="-instance-type ${PODVM_INSTANCE_TYPE} " # default t3.small
+[[ "${SSH_KP_NAME}" ]] && optionals+="-keyname ${SSH_KP_NAME} " # if not retrieved from IMDS
+[[ "${AWS_SUBNET_ID}" ]] && optionals+="-subnetid ${AWS_SUBNET_ID} " # if not set retrieved from IMDS
+[[ "${AWS_REGION}" ]] && optionals+="-aws-region ${AWS_REGION} " # if not set retrieved from IMDS
```
hmm doesn't the above changes the logic? Earlier the ami, instance etc were only needed when launch template was not used. Now it's not the case..

snir911 replied:
yes, variables passing logic has changed but caa logic is the same,  i.e. if you choose to set these ami, subnet.. vars in addition to launch template they will be passed to caa but will be ignored in favor of using the template.
I can change it but i think it makes more sense to avoid doing non-mandatory logic (which is already done in the caa code) in the entrypoint.sh, additionally, it will allow to have template fallback if we wish (in the future)

bpradipt replied:
I see.. Thanks for the explanation.

bpradipt commented on pkg/adaptor/hypervisor/aws/ec2.go:
```diff
@@ -76,3 +76,42 @@ func NewEC2Client(cloudCfg Config) (*ec2.Client, error) {
 	return client, nil
 }
+
+func RetrieveMissingConfig(cfg *Config) error {
```
nit: Does it make sense to name this method as `AddAWSConfigs` ?

snir911 replied:
maybe it does, the question is whether we should mention this configs are retrieved?

bpradipt replied:
Yeah you are right. I couldn't think of a better name other than "RetrieveMissingConfig"

bpradipt commented on pkg/adaptor/hypervisor/aws/ec2.go:
```diff
@@ -76,3 +76,42 @@ func NewEC2Client(cloudCfg Config) (*ec2.Client, error) {
 }
+
+func RetrieveMissingConfig(cfg *Config) error {
+	mdr := NewMetadataRetriever()
```
checking for mdr ?

snir911 replied:
it fails only if it cannot fetch the mac address, which is not always required and if some field does require it will fail on Get
(imds client creating also doesn't seem to fail)

bpradipt replied:
ok. May be add a comment just above this call for future use.

bpradipt commented on pkg/adaptor/hypervisor/aws/types.go:
```diff
@@ -19,3 +24,35 @@ type Config struct {
+	r.client = imds.New(imds.Options{ClientEnableState: imds.ClientDefaultEnableState}) // use imds.ClientEnabled to enforce enabling
+	mac, err := r.Get("mac")
+	if err != nil {
+		// print warn // will fail on Get
```
Better to log the error to help with debugging.

snir911 replied:
right, i missed that, thanks!

snir911 commented:
changes:
- changed function names to lower case as it should be used from within the aws pkg
- fixes as commented (note: in types.go i used to fmt as logger is not available for all build tags)

bpradipt commented on pkg/adaptor/hypervisor/aws/types.go:
```diff
@@ -19,3 +25,36 @@ type Config struct {
+}
+
+// no error return, if something is wrong, will fail on Get()
+func newMetadataRetriever() *MetadataRetriever {
```
@snir911 as discussed offline, let's move the imds methods to it's own file.

snir911 commented:
changes:
* moved imds to a new file (which allows to use logger now)
* added SGs retrieve (also adding WIP as it was not tested yet)
also, as multiple SGs are allowed and i cannot tell which one should be used by the caa, it fetches all the SGs the node is using, is this security compromise or does it make sense?

bpradipt commented:
> changes:
> 
> * moved imds to a new file (which allows to use logger now)
> * added SGs retrieve (also adding WIP as it was not tested yet)
>   also, as multiple SGs are allowed and i cannot tell which one should be used by the caa, it fetches all the SGs the node is using, is this security compromise or does it make sense?

I think it should be fine. If a specific SG is required then it should be specified in the config

snir911 commented:
tested (quay.io/snir/cloud-api-adaptor-aws:315)

bpradipt commented on cmd/cloud-api-adaptor/main.go:
```diff
@@ -57,13 +57,13 @@ func (cfg *daemonConfig) Setup() (cmd.Starter, error) {
 			flags.StringVar(&awscfg.ImageId, "imageid", "", "Pod VM ami id")
 			flags.StringVar(&awscfg.InstanceType, "instance-type", "t3.small", "Pod VM instance type")
-			flags.StringVar(&awscfg.SecurityGroupId, "securitygroupid", "", "Security Group Id to be used for the Pod VM")
+			flags.Var(&awscfg.SecurityGroupIds, "securitygroupid", "Security Group Id to be used for the Pod VM")
```
The short name be changed should also be changed to `securitygroupids`

bpradipt replied:
Another aspect is how the multiple values should be passed from the cli

snir911 replied:
Currently cmdline supports only single SG, the reason i added support to multiple SGs when getting it from IDMS is only because i couldn't tell which i should choose.
I couldn't think of a use-case where multiple values are needed for the cmdline but i'll add support it also if it just the sake of consistency

bpradipt commented on entrypoint.sh:
```diff
@@ -26,18 +26,16 @@ test_vars() {
-	optionals+="-subnetid ${AWS_SUBNET_ID} "
-fi
+[[ "${PODVM_LAUNCHTEMPLATE_NAME}" ]] && optionals+="-use-lt -aws-lt-name ${PODVM_LAUNCHTEMPLATE_NAME} " # has precedence if set
+[[ "${AWS_SG_ID}" ]] && optionals+="-securitygroupid ${AWS_SG_ID} " # MUST if template is not used
```
The short name be changed should also be changed to `securitygroupids`  include the variable AWS_SG_IDS

bpradipt commented on install/overlays/aws/kustomization.yaml:
```diff
@@ -19,18 +19,18 @@ configMapGenerator:
-  - SSH_KP_NAME="" #set
-  - AWS_SUBNET_ID="" #set
+  #- PODVM_INSTANCE_TYPE="t3.small" # caa defaults to t3.small
+  #- AWS_SG_ID="" # if not set retrieved from IMDS (NOTE: if not set it will use all node SGs)
```
AWS_SG_IDS

bpradipt reviewed (approved):
/lgtm
Thanks @snir911

huoqifeng commented on cmd/cloud-api-adaptor/main.go:
```diff
@@ -57,13 +57,13 @@ func (cfg *daemonConfig) Setup() (cmd.Starter, error) {
 			flags.StringVar(&awscfg.ImageId, "imageid", "", "Pod VM ami id")
 			flags.StringVar(&awscfg.InstanceType, "instance-type", "t3.small", "Pod VM instance type")
-			flags.StringVar(&awscfg.SecurityGroupId, "securitygroupid", "", "Security Group Id to be used for the Pod VM")
+			flags.Var(&awscfg.SecurityGroupIds, "securitygroupids", "Security Group Ids to be used for the Pod VM, comma separated")
```
Will this cause other providers code change also? comma separated array looks has less impact but seems the key changed from `securitygroupid` to `securitygroupids`.

bpradipt replied:
It won't have impact to other providers since the code is specific to aws. However the other providers can use the same logic to add support for inputs which can take multiple values.

bpradipt replied:
@huoqifeng hope this addresses your question ?

huoqifeng reviewed (approved):
LGTM

yoheiueda reviewed (approved):
LGTM
